### PR TITLE
[SYCL] Always let the backend choose the binary

### PR DESCRIPTION
**Pull request by hiaselhans**
There is still a little issue that in nvptx binaries entries are not distributed.
When compiling with `-fsycl-targets=nvptx64-nvidia-opencl-sycldevice,spir64-unknown-opencl-sycldevice` programm_manager is only providing the spir64 binary which has the proper kernel name in entries. that's why the length of binaries is 1 and sycl directly continues to execute with the wrong binary and the error thrown at a later point is misleading:

```
terminate called after throwing an instance of 'cl::sycl::feature_not_supported'
  what():  Online compilation is not supported in this context -59 (CL_INVALID_OPERATION)
```

The overhead by always having the backend check for valid binary is not too big in my eyes and the error code becomes that:

```
terminate called after throwing an instance of 'cl::sycl::runtime_error'
  what():  OpenCL API failed. OpenCL API returns: -42 (CL_INVALID_BINARY) -42 (CL_INVALID_BINARY)
```

While the issue with entry names remains (i will look into it but not sure where to start), i guess this also happened in a case where binaries are completely missing for the selected device backend.

### CODE REVIEW

**Comment by hiaselhans** [comment]
see #1588 for the related binary-entries-table issue

**Comment by Ruyk** [comment]
Just to understand this, the intention of the patch is to provide a better error message? or are you going to add more later on?

**Comment by hiaselhans** [comment]
the only intention of this PR is to prevent sending wrong binaries to the backend.

the conditional exception allowed to enter a path that should not be allowed (continuing with a non-matching backend/binary combination).

**Comment by kbobrovs** [comment]
> i will look into it but not sure where to start

sycl-post-link tool generates the symbol table - use -symbols switch.
As PTX does not support spec constants, post-link should generate 2-column table - first column lists binaries, second - corresponding symbol files. See Driver.cpp/appendLinkDependences code after  the picture illustrating the action graph.  To support symbol table, post-link should generate TY_Tempfiletable in case of PTX too. Use file-table-tform to process filenames in the table, similar to non-PTX path

**Comment by hiaselhans** [comment]
So, with this PR there is a failing test: `kernel_from_file`

```
>>> ProgramManager::getDeviceImage(-1, "0", 0x89d880)
available device images:
  ++++++ Kernel set: 0
  --- Image 0xd934c0
    Version  : 1
    Kind     : 4
    Format   : 2
    Target   : <unknown>
    Bin size : 13532
    Compile options : 
    Link options    : 
    Entries  : 
    Properties [0-0]:
    OSModuleHandle=-2
    DYNAMICALLY CREATED
```

programm_manager fails to select the image because target is set to unknown via [DynRTDeviceBinaryImage](https://github.com/intel/llvm/blob/03208c000d28a6f2234bcd0f3f9122d9f900a0cd/sycl/source/detail/program_manager/program_manager.cpp#L990)

for now DynRTDeviceBinaryImage is only used within the `UseSpvEnv` so we could hardcode the target to spir64?

**Comment by kbobrovs** [comment]
> for now DynRTDeviceBinaryImage is only used within the UseSpvEnv so we could hardcode the target to spir64?

The SPIRV format should have been determined by`Bin->Format = pi::getBinaryImageFormat(Bin->BinaryStart, DataSize);` in the constructor. I think this should be investigated/fixed rather than hardcoding SPIRV.

**Comment by kbobrovs** [comment]
OK, the problem is in target, not in the format. So I'd suggest then to initialize Bin->target with SPIRV64 depending on Bin->format.

**Comment by hiaselhans** [comment]
> OK, the problem is in target, not in the format. So I'd suggest then to initialize Bin->target with SPIRV64 depending on Bin->format.

yep, i just found that. so i will set target with a switch statement checking format:

pi.h:
```
static constexpr pi_device_binary_type PI_DEVICE_BINARY_TYPE_NONE = 0;
// specific to a device
static constexpr pi_device_binary_type PI_DEVICE_BINARY_TYPE_NATIVE = 1;
// portable binary types go next
// SPIR-V
static constexpr pi_device_binary_type PI_DEVICE_BINARY_TYPE_SPIRV = 2;
// LLVM bitcode
static constexpr pi_device_binary_type PI_DEVICE_BINARY_TYPE_LLVMIR_BITCODE = 3;
```

**Comment by kbobrovs** [comment]
yes, sounds good

**Comment by hiaselhans** [comment]
Did that, it doesn't feel a little hard-coded still but it's a bit of an edge case anyways.

thx @kbobrovs !

File: sycl/source/detail/program_manager/program_manager.cpp [hunk 1]
```diff
@@ -665,9 +665,13 @@ void ProgramManager::populateSpecConstRegistry() {
 RTDeviceBinaryImage &ProgramManager::getDeviceImage(OSModuleHandle M,
                                                     KernelSetId KSId,
                                                     const context &Context) {
-  if (DbgProgMgr > 0)
+  if (DbgProgMgr > 0) {
     std::cerr << ">>> ProgramManager::getDeviceImage(" << M << ", \"" << KSId
               << "\", " << getRawSyclObjImpl(Context) << ")\n";
+
+    std::cerr << "available device images:\n";
```
Comment (Ruyk) [inline sycl/source/detail/program_manager/program_manager.cpp:672]: Could help when filtering the output...

```suggestion
    std::cerr << "ProgramManager: Available device images:\n";
```

File: sycl/source/detail/program_manager/program_manager.cpp [hunk 2]
```diff
@@ -1004,6 +1003,15 @@ DynRTDeviceBinaryImage::DynRTDeviceBinaryImage(
...
   Bin->EntriesEnd = nullptr;
   Bin->Format = pi::getBinaryImageFormat(Bin->BinaryStart, DataSize);
+
+  switch (Bin->Format) {
+  case PI_DEVICE_BINARY_TYPE_SPIRV:
+    Bin->DeviceTargetSpec = PI_DEVICE_BINARY_TARGET_SPIRV64;
+    break;
+  default:
+    Bin->DeviceTargetSpec = PI_DEVICE_BINARY_TARGET_UNKNOWN;
+  }
```
Comment (Ruyk) [inline sycl/source/detail/program_manager/program_manager.cpp:1013]: Does this work with the CUDA backend when there are multiple binaries?

Comment (hiaselhans) [reply sycl/source/detail/program_manager/program_manager.cpp:1013]: i think you can only load one `DynRTDeviceBinaryImage` at a time using the `SYCL_USE_KERNEL_SPV` env-variable. this just sets the devicetarget to spir64 in case it is a spir-file so the opencl backend can recognize it.

File: sycl/source/detail/program_manager/program_manager.cpp (hunk 1 above)
Comment (hiaselhans) [reply sycl/source/detail/program_manager/program_manager.cpp:672]: actually i just moved this part up because it might be unreachable at it's old place.

If i were to improve those messages i would query for PI_TRACE at runtime instead? what do you think?

**Comment by hiaselhans** [comment]
> Just few comments, I haven't tried with the CUDA backend, does this patch solves the problem?

It doesn't yet solve the issue quoted, but it does help when someone compiles only nvptx64 binaries and uses an opencl device and vice-versa. The if-clause prevented the backend from checking the image type before executing it.

**Comment by Ruyk** [comment]
https://github.com/intel/llvm/pull/1543 rejects OpenCL CUDA platform on DPC++, so should help as well.

**Comment by hiaselhans** [comment]
from looking at #1543 i do like the fact that there's now always `getBackend() == backend::opencl`
makes a lot of sense to me! :)

however there's still this: isDeviceBinaryTypeSupported
https://github.com/intel/llvm/blob/ec0846c44910d4b15f1c8b22207a61b20c2c0ac5/sycl/source/detail/program_manager/program_manager.cpp#L259
so i wonder if we could have a PI call `piDeviceSupported(pi_device *device, bool *support)` and reject all those opencl 1.2 devices altogether?

File: sycl/source/detail/program_manager/program_manager.cpp (hunk 1 above)
Comment (Ruyk) [reply sycl/source/detail/program_manager/program_manager.cpp:672]: That would be even better

**Comment by Ruyk** [comment]
I don't think we should ban OpenCL 1.2 devices, since lots of them could work (specially if they expose `cl_khr_il_program`). This is checked below in 

https://github.com/intel/llvm/blob/ec0846c44910d4b15f1c8b22207a61b20c2c0ac5/sycl/source/detail/program_manager/program_manager.cpp#L279-L290

It is also, IMHO,  not a problem of the PI API to decide if a device is valid for a plugin, one of the reason is because the `pi_device` object may come from a different object to the plugin you are calling the function and they are not necessarily compatible (there is no guarantee on PI API that PI types are compatible at all across implementations). 
Since a `pi_device` is returned from querying a  `pi_platform`  , by definition, the `pi_device` should be valid.

In terms of CUDA, note the `PiDeviceBinaryType` check:
https://github.com/intel/llvm/blob/ec0846c44910d4b15f1c8b22207a61b20c2c0ac5/sycl/source/detail/program_manager/program_manager.cpp#L262-L263, 

Currently, a PTX binary is appearing as NATIVE, so this function itself is not used,

https://github.com/intel/llvm/blob/ec0846c44910d4b15f1c8b22207a61b20c2c0ac5/sycl/source/detail/program_manager/program_manager.cpp#L335-L338 

The code is handled as a Binary blob loaded, like any other native binary format:

https://github.com/intel/llvm/blob/ec0846c44910d4b15f1c8b22207a61b20c2c0ac5/sycl/source/detail/program_manager/program_manager.cpp#L345-L348


FYI, definition of Binary types:
https://github.com/intel/llvm/blob/ec0846c44910d4b15f1c8b22207a61b20c2c0ac5/sycl/include/CL/sycl/detail/pi.h#L579-L587

**Comment by hiaselhans** [comment]
thx @Ruyk 

Ok, i see. I was wrong in a way that `isDeviceBinaryTypeSupported` does return true on opencl >= 2.1 and not false <=. Wrong assumption from my side, but in a way i just meant if we could use that logic there to filter unsupported opencl devices.

So let me rephrase my point:

The isDeviceBinaryTypeSuported function does **only** check for format=PI_DEVICE_BINARY_TYPE_SPIRV if a device supports sprir-v.
It is the opencl backend to deliver these pi_devices, so why shouldn't it also be responsible to check if that device is a "good" opencl device? in a way it could also be a bool `isSupported` or similar stored with pi_device which can be checked in device_selector.

File: sycl/source/detail/program_manager/program_manager.cpp (hunk 1 above)
Comment (hiaselhans) [reply sycl/source/detail/program_manager/program_manager.cpp:672]: @kbobrovs should i replace all `DbgProgMgr` checks with pi_trace in programmmanager?

**Comment by Ruyk** [comment]
> It is the opencl backend to deliver these pi_devices, so why shouldn't it also be responsible to check if that device is a "good" opencl device? in a way it could also be a bool `isSupported` or similar stored with pi_device which can be checked in device_selector.

Any OpenCL device is a "good" opencl device, it is only the SYCL RT that can use or not a certain OpenCL (or PI device). PI Plugin should provide devices and platforms to the SYCL RT, decisions on what is valid is something that should be on the SYCL RT layers.

**Comment by hiaselhans** [comment]
i think this discussion should be in #1543 because nothing of it is actually addressed here, in this pr.

in other words: If all devices are "good" devices, why do we need special treatment for the cuda opencl device? why not reusing the logic from `isDeviceBinaryTypeSupported` and as that logic is quite opencl specific, why not moving it to the opencl backend?

forgive my naive questions but it just seems so much more straightforward to me...

**Comment by bader** [comment]
@hiaselhans, sorry for the delay.
It looks like there is a conflict with https://github.com/intel/llvm/commit/c22e34b39ad2ab38f885b2313a4f5a22dd7b0553. Could you resolve it, please?

**Comment by hiaselhans** [comment]
@bader  done :)

File: sycl/source/detail/program_manager/program_manager.cpp (hunk 1 above)
Comment (kbobrovs) [reply sycl/source/detail/program_manager/program_manager.cpp:644]: @hiaselhans , sorry for delay. Yes, definitely makes sense. Can be done as a separate PR.

**Comment by bader** [comment]
@intel/llvm-reviewers-runtime, ping.
